src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -200,41 +234,107 @@ uct_cuda_ipc_open_memhandle_vmm(uct_cuda_ipc_rkey_t *key,
     return status;
 }
 
-static ucs_status_t
-uct_cuda_ipc_open_memhandle_mempool(uct_cuda_ipc_rkey_t *key,
-                                    CUdeviceptr *mapped_addr)
+static ucs_status_t cuda_ipc_rem_mpool_cache_create(uct_cuda_ipc_rkey_t *key,
+                                                    CUmemoryPool **mpool,
+                                                    CUdeviceptr *mapped_addr)
 {
     CUmemAccessDesc access_desc = {};
-    CUdeviceptr dptr;
     ucs_status_t status;
 
+    *mpool = ucs_malloc(sizeof(CUmemoryPool), "cuda_ipc_mempool");
+    if (*mpool == NULL) {
+        ucs_error("failed to allocate memory for locally mapped mempool");
+        return UCS_ERR_NO_MEMORY;
+    }
+
     status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportFromShareableHandle(
-                &key->ph.pool, (void *)&key->ph.handle.fabric_handle,
+                *mpool, (void *)&key->ph.handle.fabric_handle,
                 CU_MEM_HANDLE_TYPE_FABRIC, 0));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(&dptr,
-                key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                *(*mpool), (CUmemPoolPtrExportData*)&key->ph.ptr));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
     status = uct_cuda_ipc_init_access_desc(&access_desc);
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolSetAccess(key->ph.pool,
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolSetAccess(*(*mpool),
                 &access_desc, 1));
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
-    *mapped_addr = dptr;
-
     return UCS_OK;
+
+err:
+    ucs_free(*mpool);
+    *mpool = NULL;
+    return status;
+}
+
+static ucs_status_t
+uct_cuda_ipc_open_memhandle_mempool(uct_cuda_ipc_rkey_t *key,
+                                    CUdeviceptr *mapped_addr)
+{
+    khash_t(cuda_ipc_rem_mpool_cache) *hash = &uct_cuda_ipc_rem_mpool_cache.hash;
+    CUmemFabricHandle hkey                  = key->ph.handle.fabric_handle;
+    ucs_status_t status                     = UCS_OK;
+    CUmemoryPool *mpool;
+    khiter_t khiter;
+    int khret;
+
+    pthread_rwlock_rdlock(&uct_cuda_ipc_rem_mpool_cache.lock);
+    khiter = kh_get(cuda_ipc_rem_mpool_cache, hash, hkey);
+    if (ucs_likely(khiter != kh_end(hash))) {
+        key->ph.pool = kh_val(hash, khiter);
+        ucs_assert(key->ph.pool != NULL);
+
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemPoolImportPointer(mapped_addr,
+                    key->ph.pool, (CUmemPoolPtrExportData*)&key->ph.ptr));
+        goto err;
```

Comment:
do you think it is performance-wise reasonable trying to `kh_get` at first with rd lock?
You could just do put immediately and take existing value if `(khret == UCS_KH_PUT_KEY_PRESENT)`
This would be a bit more intrusive, but code will be simpler

Reply:
Let me respond here, because it was me who suggested to use rwlock.
The aim of using rwlock here was to minimise the contention, so that multiple readers can progress simultaneously without blocking each other at all (when cache is populated). Typically it's desirable to have less global contention points. I don't think it complicates a lot

Reply:
ok